cmd/notifier/messenger.go

```diff
@@ -0,0 +1,135 @@
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"log"
+	"time"
+
+	"github.com/status-im/status-go/geth/node"
+	"github.com/status-im/status-go/messenger"
+)
+
+const (
+	// defaultWorkTime is a work time reported in messages sent to MailServer nodes.
+	defaultWorkTime = 5
+)
+
+var (
+	// ErrInvalidSymKeyID is returned when it fails to get a symmetric key.
+	ErrInvalidSymKeyID = errors.New("invalid symKeyID value")
+)
+
+type BroadcastMsg struct {
+	PubKey string `json:"pubkey"`
+}
+
+// RegistrationRequestMsg : peers wanting to use this notification server will
+// send anonymous whisper messages with the device registration token, and
+// a secure channel, so communication between PN server and client can happen
+// securely
+type RegistrationRequestMsg struct {
+	DeviceRegistrationToken string `json:"token"`
+	SecureChannel           []byte `json:"channel"`
+}
+
+// Messenger : whisper interface for the notifier
+type Messenger struct {
+	symKey         []byte
+	discoveryTopic string
+	pollInterval   time.Duration
+	addressKey     string
+	password       string
+	client         *messenger.Messenger
+}
+
+// NotificationRequestMsg : a registered contact requests sending a push
+// notification to one of its contacts
+type NotificationRequestMsg struct {
+	// TODO (adriacidre) : Check @PombeirP what fields are needed here
+}
+
+// NewMessenger Creates a new Messenger
+func NewMessenger(sn *node.StatusNode, discoveryTopic string, pollInterval time.Duration) *Messenger {
+	password := "password"
+	client := messenger.New(sn)
+	addr := client.Signup(password)
+	log.Println("ADDRESS :", addr)
+	addressKey, err := client.Login(addr, password)
+	log.Println("ADDRESS KEY :", addr)
+	if err != nil {
+		log.Println("An error logging in")
+	}
+
+	return &Messenger{
+		password:       password,
+		addressKey:     addressKey,
+		pollInterval:   pollInterval,
+		discoveryTopic: discoveryTopic,
+		client:         client,
+	}
+}
+
+// BroadcastAvailability : Broadcasts its availability to serve as
+// notification server
+func (m *Messenger) BroadcastAvailability() error {
+	_, _, channelKey := m.client.JoinPublicChannel(m.discoveryTopic)
+
+	go func() {
+		for range time.Tick(m.pollInterval) {
+			log.Println("Broadcasting availability on", m.discoveryTopic)
+			m.client.NewContactKeyRequest(m.addressKey, m.discoveryTopic, channelKey, "me")
+		}
+	}()
+
+	return nil
+}
+
+// ManageRegistrations clients will be sending registration requests to the
+// messenger topic, this method retrieves those messages and stores its
+// information to allow push notifications
+func (m *Messenger) ManageRegistrations() {
+	log.Println("Subscribed to discovery topic :", m.discoveryTopic)
+	filterID, _, _ := m.client.JoinPublicChannel(m.discoveryTopic)
+	for range time.Tick(m.pollInterval) {
```

Comment:
Yes, I think that's how it works, and its a pita in terms of performance (as we will have n topic to be subscribed).

On the other hand I think you can subscribe to multiple topics at the same time, so we can allow subscription object to manage the topics is subscribed to.

Reply:
From the [source code](https://github.com/status-im/status-go/blob/feature/issue-847/gorush-based-push-notification-server/messenger/subscription.go#L33) it looks like it will wait for messages indefinitely, so I don't see why we need to call `Subscribe` over and over again with the same `filterID`...

Reply:
Ah, damn it, you're right. Subscribe is already implementing this loop, it doesn't need to exist here